File: backend/scripts/seed.ts
#!/usr/bin/env tsx
/**
 * Seed canonical tables from db.xlsx.
 *
 * Reads: backend/data/db.xlsx  (or XLSX_PATH env var)
 * Truncates and reseeds: product, supplier, purchase_order,
 *   purchase_order_line, supplier_product, supplier_email_aliases
 *
 * Safe to run multiple times in development.
 */

import pg from 'pg'
import XLSX from 'xlsx'

const { Pool } = pg
const pool = new Pool({ connectionString: process.env.DATABASE_URL })

// ── Excel date conversion ────────────────────────────────────────────────────
// Excel stores dates as days since 1899-12-30 (the "1900 date system").
const EXCEL_EPOCH_MS = new Date(Date.UTC(1899, 11, 30)).getTime()

function excelDateToISO(val: unknown): string | null {
  if (val === null || val === undefined || val === '') return null
  if (typeof val !== 'number') return null
  return new Date(EXCEL_EPOCH_MS + val * 86_400_000).toISOString().slice(0, 10)
}

function nullify(val: unknown): unknown {
  if (val === null || val === undefined || val === '' || val === '-') return null
  return val
}

// ── main ─────────────────────────────────────────────────────────────────────

async function seed() {
  const xlsxPath = process.env.XLSX_PATH ?? './backend/data/db.xlsx'
  const wb = XLSX.readFile(xlsxPath)

  const read = (sheet: string) =>
    XLSX.utils.sheet_to_json<Record<string, unknown>>(wb.Sheets[sheet], {
      defval: null,
    })

  const productRows   = read('product')
  const supplierRows  = read('supplier')
  const poRows        = read('purchase_order')
  const polRows       = read('purchase_order_line')
  const spRows        = read('supplier_product')

  console.log(
    `xlsx loaded: ${productRows.length} products, ${supplierRows.length} suppliers,` +
    ` ${poRows.length} POs, ${polRows.length} lines, ${spRows.length} supplier_products`,
  )

  const db = await pool.connect()
  try {
    await db.query('BEGIN')

    // Truncate in reverse FK dependency order; CASCADE handles child rows.
    await db.query(
      `TRUNCATE supplier_email_aliases, supplier_product,
                purchase_order_line, purchase_order,
                product, supplier
       RESTART IDENTITY CASCADE`,
    )

    // Legacy-id → new uuid maps for FK resolution
    const productMap  = new Map<number, string>()
    const supplierMap = new Map<number, string>()
    const poMap       = new Map<number, string>()

    // ── product ──────────────────────────────────────────────────────────────
    for (const r of productRows) {
      const { rows } = await db.query(
        `INSERT INTO product (legacy_id, sku, title) VALUES ($1, $2, $3) RETURNING id`,
        [r.id, r.sku, nullify(r.title)],
      )
      productMap.set(r.id as number, rows[0].id)
    }
    console.log(`  ✓ ${productRows.length} products`)

    // ── supplier + one alias per supplier ────────────────────────────────────
    for (const r of supplierRows) {
      const { rows } = await db.query(
        `INSERT INTO supplier (legacy_id, name, email) VALUES ($1, $2, $3) RETURNING id`,
        [r.id, r.name, r.email],
      )
      const newId = rows[0].id as string
      supplierMap.set(r.id as number, newId)

      // Seed the primary email as an alias so alias-based resolution works too.
      await db.query(
        `INSERT INTO supplier_email_aliases (supplier_id, email_address) VALUES ($1, $2)`,
        [newId, r.email],
      )
    }
    console.log(`  ✓ ${supplierRows.length} suppliers (+ ${supplierRows.length} email aliases)`)

    // ── purchase_order ───────────────────────────────────────────────────────
    for (const r of poRows) {
      const supplierId = supplierMap.get(r.supplier_id as number)
      if (!supplierId) throw new Error(`Unknown supplier legacy_id: ${r.supplier_id}`)
      const { rows } = await db.query(
        `INSERT INTO purchase_order (legacy_id, reference_num, supplier_id, delivery_date)
         VALUES ($1, $2, $3, $4) RETURNING id`,
        [r.id, r.reference_num, supplierId, excelDateToISO(r.delivery_date)],
      )
      poMap.set(r.id as number, rows[0].id)
    }
    console.log(`  ✓ ${poRows.length} purchase_orders`)

    // ── purchase_order_line (batch in chunks of 100) ─────────────────────────
    const polParams: unknown[][] = polRows.map((r) => {
      const poId      = poMap.get(r.purchase_order_id as number)
      const productId = productMap.get(r.product_id as number)
      if (!poId)      throw new Error(`Unknown PO legacy_id: ${r.purchase_order_id}`)
      if (!productId) throw new Error(`Unknown product legacy_id: ${r.product_id}`)
      return [r.id, poId, productId, r.quantity, excelDateToISO(r.delivery_date)]
    })

    const CHUNK = 100
    for (let i = 0; i < polParams.length; i += CHUNK) {
      const chunk = polParams.slice(i, i + CHUNK)
      const placeholders = chunk
        .map((_, j) => {
          const b = j * 5
          return `($${b + 1},$${b + 2},$${b + 3},$${b + 4},$${b + 5})`
        })
        .join(',')
      await db.query(
        `INSERT INTO purchase_order_line
           (legacy_id, purchase_order_id, product_id, quantity, delivery_date)
         VALUES ${placeholders}`,
        chunk.flat(),
      )
    }
    console.log(`  ✓ ${polRows.length} purchase_order_lines`)

    // ── supplier_product ─────────────────────────────────────────────────────
    for (const r of spRows) {
      const supplierId = supplierMap.get(r.supplier_id as number)
      const productId  = productMap.get(r.product_id as number)
      if (!supplierId || !productId)
        throw new Error(`Unknown ids in supplier_product row: ${JSON.stringify(r)}`)
      await db.query(
        `INSERT INTO supplier_product (supplier_id, product_id, supplier_sku, price_per_unit)
         VALUES ($1, $2, $3, $4)`,
        [supplierId, productId, nullify(r.supplier_sku), nullify(r.price_per_unit)],
      )
    }
    console.log(`  ✓ ${spRows.length} supplier_products`)

    await db.query('COMMIT')
    console.log('Seed complete.')
  } catch (err) {
    await db.query('ROLLBACK').catch(() => {})
    throw err
  } finally {
    db.release()
    await pool.end()
  }
}

seed().catch((e) => {
  console.error(e)
  process.exit(1)
})

File: backend/scripts/seed.py
"""Reset the database to the demo state.

Two steps, so that one command gives a reviewer something to look at:

  1. Reload the canonical tables from backend/data/db.xlsx, mapping the
     workbook's integer ids onto UUIDs.
  2. Ingest the committed fixture emails through the real pipeline, which
     populates the review queue.

This is destructive by design — it truncates the pipeline tables too, so
running it twice produces exactly the same state rather than stacking a
second set of proposed changes on top of the first. It is a development and
demo command; nothing calls it at runtime.

    python -m backend.scripts.seed              # canonical data + fixtures
    python -m backend.scripts.seed --no-fixtures
"""

from __future__ import annotations

import argparse
import asyncio
import datetime as dt
import os
from pathlib import Path
from typing import Any

from openpyxl import load_workbook

from backend.config import PROJECT_ROOT
from backend.db import close_pool, connection, open_pool


def _rows(worksheet: Any) -> list[dict[str, Any]]:
    """Read a sheet as dicts keyed by its header row, skipping blank rows."""
    it = worksheet.iter_rows(values_only=True)
    try:
        header = [str(h) if h is not None else "" for h in next(it)]
    except StopIteration:
        return []

    out: list[dict[str, Any]] = []
    for row in it:
        if all(cell is None for cell in row):
            continue
        out.append({key: value for key, value in zip(header, row) if key})
    return out


def _int(value: Any) -> int | None:
    """Workbook integers arrive as floats (1.0); normalise them."""
    return None if value is None else int(value)


def _date(value: Any) -> dt.date | None:
    if value is None or value == "":
        return None
    if isinstance(value, dt.datetime):
        return value.date()
    if isinstance(value, dt.date):
        return value
    return None


def _nullify(value: Any) -> Any:
    """The workbook uses '' and '-' to mean "no value"."""
    if value is None or value == "" or value == "-":
        return None
    return value


async def seed() -> None:
    xlsx_path = Path(
        os.environ.get("XLSX_PATH", PROJECT_ROOT / "backend" / "data" / "db.xlsx")
    )
    workbook = load_workbook(xlsx_path, read_only=True, data_only=True)

    products = _rows(workbook["product"])
    suppliers = _rows(workbook["supplier"])
    purchase_orders = _rows(workbook["purchase_order"])
    lines = _rows(workbook["purchase_order_line"])
    supplier_products = _rows(workbook["supplier_product"])
    workbook.close()

    print(
        f"xlsx loaded: {len(products)} products, {len(suppliers)} suppliers, "
        f"{len(purchase_orders)} POs, {len(lines)} lines, "
        f"{len(supplier_products)} supplier_products"
    )

    async with connection() as conn:
        async with conn.transaction():
            # Truncate in reverse FK dependency order; CASCADE handles children.
            # The pipeline tables go too: proposed_changes reference canonical
            # rows by a logical FK that Postgres cannot cascade, so leaving
            # them behind would strand them against ids that no longer exist —
            # and re-running the fixtures would stack a duplicate queue.
            await conn.execute(
                """
                TRUNCATE audit_log, proposed_changes, extraction_runs,
                         email_attachments, emails, supplier_corrections,
                         supplier_email_aliases, supplier_product,
                         purchase_order_line, purchase_order,
                         product, supplier
                RESTART IDENTITY CASCADE
                """
            )

            product_ids: dict[int, str] = {}
            supplier_ids: dict[int, str] = {}
            po_ids: dict[int, str] = {}

            for row in products:
                cur = await conn.execute(
                    "INSERT INTO product (legacy_id, sku, title) VALUES (%s,%s,%s) RETURNING id",
                    (_int(row["id"]), row["sku"], _nullify(row["title"])),
                )
                product_ids[_int(row["id"])] = (await cur.fetchone())["id"]
            print(f"  ✓ {len(products)} products")

            for row in suppliers:
                cur = await conn.execute(
                    "INSERT INTO supplier (legacy_id, name, email) VALUES (%s,%s,%s) RETURNING id",
                    (_int(row["id"]), row["name"], row["email"]),
                )
                supplier_id = (await cur.fetchone())["id"]
                supplier_ids[_int(row["id"])] = supplier_id

                # Seed the primary email as an alias so alias resolution works too.
                await conn.execute(
                    "INSERT INTO supplier_email_aliases (supplier_id, email_address) VALUES (%s,%s)",
                    (supplier_id, row["email"]),
                )
            print(f"  ✓ {len(suppliers)} suppliers (+ {len(suppliers)} email aliases)")

            for row in purchase_orders:
                supplier_id = supplier_ids.get(_int(row["supplier_id"]))
                if supplier_id is None:
                    raise ValueError(f"Unknown supplier legacy_id: {row['supplier_id']}")
                cur = await conn.execute(
                    """
                    INSERT INTO purchase_order
                      (legacy_id, reference_num, supplier_id, delivery_date)
                    VALUES (%s,%s,%s,%s) RETURNING id
                    """,
                    (
                        _int(row["id"]),
                        row["reference_num"],
                        supplier_id,
                        _date(row["delivery_date"]),
                    ),
                )
                po_ids[_int(row["id"])] = (await cur.fetchone())["id"]
            print(f"  ✓ {len(purchase_orders)} purchase_orders")

            for row in lines:
                po_id = po_ids.get(_int(row["purchase_order_id"]))
                product_id = product_ids.get(_int(row["product_id"]))
                if po_id is None:
                    raise ValueError(f"Unknown PO legacy_id: {row['purchase_order_id']}")
                if product_id is None:
                    raise ValueError(f"Unknown product legacy_id: {row['product_id']}")
                await conn.execute(
                    """
                    INSERT INTO purchase_order_line
                      (legacy_id, purchase_order_id, product_id, quantity, delivery_date)
                    VALUES (%s,%s,%s,%s,%s)
                    """,
                    (
                        _int(row["id"]),
                        po_id,
                        product_id,
                        row["quantity"],
                        _date(row["delivery_date"]),
                    ),
                )
            print(f"  ✓ {len(lines)} purchase_order_lines")

            for row in supplier_products:
                supplier_id = supplier_ids.get(_int(row["supplier_id"]))
                product_id = product_ids.get(_int(row["product_id"]))
                if supplier_id is None or product_id is None:
                    raise ValueError(f"Unknown ids in supplier_product row: {row}")
                await conn.execute(
                    """
                    INSERT INTO supplier_product
                      (supplier_id, product_id, supplier_sku, price_per_unit)
                    VALUES (%s,%s,%s,%s)
                    """,
                    (
                        supplier_id,
                        product_id,
                        _nullify(row["supplier_sku"]),
                        _nullify(row["price_per_unit"]),
                    ),
                )
            print(f"  ✓ {len(supplier_products)} supplier_products")

    print("Canonical tables reloaded.")


async def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--no-fixtures",
        action="store_true",
        help="load canonical data only, leaving the review queue empty",
    )
    args = parser.parse_args()

    await open_pool()
    try:
        await seed()
        if not args.no_fixtures:
            from backend.scripts.load_fixtures import load_fixtures

            print("Ingesting fixture emails through the pipeline...")
            totals = await load_fixtures()
            print(
                f"  {totals['proposed']} changes proposed, "
                f"{totals['auto_applied']} auto-applied, "
                f"{totals['pending']} awaiting review."
            )
    finally:
        await close_pool()


if __name__ == "__main__":
    asyncio.run(main())
